druid/src/widget/svg.rs

```diff
@@ -0,0 +1,230 @@
+// Copyright 2019 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! An SVG widget.
+
+use crate::{
+    kurbo::BezPath,
+    piet::{Color, RenderContext},
+    BaseState, BoxConstraints, Data, Env, Event, EventCtx, LayoutCtx, PaintCtx, Point, Size,
+    UpdateCtx, Widget,
+};
+
+use log::error;
+use std::error::Error;
+use std::marker::PhantomData;
+use std::str::FromStr;
+
+use usvg;
+
+/// A widget that renders a SVG
+pub struct Svg<T> {
+    // On construction the SVG string is parsed into a SvgData.
+    svg_data: SvgData,
+    phantom: PhantomData<T>,
+}
+
+impl<T: Data> Svg<T> {
+    /// Create an SVG-drawing widget from a valid SVG string literal.
+    ///
+    /// The SVG will scale to fit its box constraints.
+    /// If SVG is invalid a blank SVG will be rendered instead.
+    pub fn new(svg_str: &str) -> impl Widget<T> {
```

Comment:
If this took `SvgData` rather than `str`, then redundant work could be avoided in constructing multiple widgets with the same SVG, and there'd be no need to deal with errors internally.

Reply:
This might be a good compromise around the error handling; you are expected to check errors when constructing `SvgData`, but then you're free to do what you want.

We could also consider implementing `Default` for `SvgData`, which would do what `Svg::empty()` does? I'm not sure if this is helpful or not.

If we expect `SvgData` to be shared, it might be worth `Arc`ing the tree?

Reply:
Okay how does this look?